### ISSUE TITLE
e2e consistent failure on GKE: Node reboot test 

### ISSUE THREAD

**Issue by quinton-hoole** [GoogleCloudPlatform/kubernetes#8378]
```
[It] should reboot each node and ensure they function upon restart
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/reboot.go:97
INFO: Getting k8s-jenkins-gke-e2e-ci2-node-1
INFO: Getting k8s-jenkins-gke-e2e-ci2-node-2
INFO: Waiting up to 20s for node k8s-jenkins-gke-e2e-ci2-node-1 readiness to be true
INFO: Waiting up to 20s for node k8s-jenkins-gke-e2e-ci2-node-2 readiness to be true
INFO: Node k8s-jenkins-gke-e2e-ci2-node-1 condition 1/1: type: Ready, status: True
INFO: Successfully found node k8s-jenkins-gke-e2e-ci2-node-1 readiness to be true
INFO: Node k8s-jenkins-gke-e2e-ci2-node-2 condition 1/1: type: Ready, status: True
INFO: Successfully found node k8s-jenkins-gke-e2e-ci2-node-2 readiness to be true
INFO: Node k8s-jenkins-gke-e2e-ci2-node-1 has 1 pods: [fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-1]
INFO: Waiting up to 20s for the following pods to be running and ready: [fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-1]
INFO: Waiting up to 20s for pod fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-1 status to be running and ready
INFO: Node k8s-jenkins-gke-e2e-ci2-node-2 has 2 pods: [fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-2 kube-dns-bktfv]
INFO: Waiting up to 20s for the following pods to be running and ready: [fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-2 kube-dns-bktfv]
INFO: Waiting up to 20s for pod fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-2 status to be running and ready
INFO: Waiting up to 20s for pod kube-dns-bktfv status to be running and ready
INFO: Waiting for pod fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-1 in namespace default status to be "running and ready" (found "Running") (3.947432ms)
INFO: Waiting for pod fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-2 in namespace default status to be "running and ready" (found "Running") (3.477613ms)
INFO: Waiting for pod fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-1 in namespace default status to be "running and ready" (found "Running") (5.010109742s)
INFO: Waiting for pod fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-2 in namespace default status to be "running and ready" (found "Running") (5.007668281s)
INFO: Waiting for pod fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-1 in namespace default status to be "running and ready" (found "Running") (10.017226337s)
INFO: Waiting for pod fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-2 in namespace default status to be "running and ready" (found "Running") (10.014183556s)
INFO: Waiting for pod fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-2 in namespace default status to be "running and ready" (found "Running") (15.020352799s)
INFO: Waiting for pod fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-1 in namespace default status to be "running and ready" (found "Running") (15.079531808s)
INFO: Pod kube-dns-bktfv failed to be running and ready.
INFO: Wanted all pods to be running and ready. Result: false. Pods: [fluentd-cloud-logging-k8s-jenkins-gke-e2e-ci2-node-2 kube-dns-bktfv]

• Failure [20.285 seconds]
Reboot
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/reboot.go:98
  should reboot each node and ensure they function upon restart [It]
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/reboot.go:97

  Node k8s-jenkins-gke-e2e-ci2-node-1 failed reboot test.
```

**Comment by quinton-hoole** [comment]
See also #8369

**Comment by roberthbailey** [comment]
Assigning to Max.

**Comment by mbforbes** [comment]
The problem happens before the node ever reboots. As a precondition sanity check, it ensures all pods are "running" and "ready" before the test begins. This isn't true, so it bails.

Before the e2e tests run, we're checking that all pods are "running" by ensuring:
`pod.Status.Phase == api.PodRunning`

This test also checks that pods are "ready" by ensuring:
`pod.Status.Conditions ⊇ {Type: api.PodReady, Status: api.ConditionTrue}`

The second check is the one that fails. It's failing for `fluentd-cloud-logging-<node>` pods. 

There are a few things we can do, ranging from "easiest" to "actually understanding what is going on:"

0. Increase this test's precondition timeout
0. Make the e2e pre-checks also ensure the pods are "ready"
0. Figure out why `fluentd-cloud-logging-<node>` pods are coming up as "running" but not "ready" on GKE

I'll start on the first two though might need advice to look into the third.

**Comment by quinton-hoole** [comment]
We're actually seeing flaky results from the cloud logging e2e tests too, across most test environments, which might be related.

Here's a recent example: 

------------------------------
Cluster level logging using Elasticsearch 
  should check that logs from pods on all nodes are ingested into Elasticsearch
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:46
[BeforeEach] Cluster level logging using Elasticsearch
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:42
>>> testContext.KubeConfig: /var/lib/jenkins/jobs/kubernetes-e2e-gce/workspace/.kube/config
[It] should check that logs from pods on all nodes are ingested into Elasticsearch
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:46
STEP: Checking the Elasticsearch service exists.
STEP: Checking to make sure the Elasticsearch pods are running
INFO: Waiting up to 5m0s for pod elasticsearch-logging-sxbih status to be running
INFO: Waiting up to 5m0s for pod elasticsearch-logging-znhy6 status to be running
STEP: Checking to make sure we are talking to an Elasticsearch service.

• Failure [0.730 seconds]
Cluster level logging using Elasticsearch
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:47
  should check that logs from pods on all nodes are ingested into Elasticsearch [It]
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:46

  Elasticsearch cluster has a bad status: 503

**Comment by satnam6502** [comment]
I merged a logging related PR #8374 this morning from @jimmidyson  -- let me investigate.

**Comment by satnam6502** [comment]
I have a PR but I can't seem to push it to GitHub. Waiting...

**Comment by mbforbes** [comment]
Update: On the GKE cluster I was testing on, the `fluentd-cloud-logging-<node>` pods are permanently   "running" but not "ready," which means #8442 won't fix this (though it's still good to have it). We need to get those pods "ready."

**Comment by mbforbes** [comment]
So, um, given this isn't the reboot test being broken, but rather the reboot test happening to catch broken pods that no other test is checking, should we re-assign this issue so that someone who knows more about logging than I do is working on it?

We can also turn off the reboot test for GKE if we'd rather have it be green (though for me, DNS and ServiceAccounts are also flaking, even when run before the reboot happens). I guess I'm unsure what the policy should be here when a test catches bugs in other code.

**Comment by satnam6502** [comment]
I still don't understand what is going on. Perhaps someone can explain things to me in RL on Tuesday.

**Comment by mbforbes** [comment]
@satnam6502 let me try really quickly, though happy to chat tomorrow.

The prerequisite to understanding this is that we normally just check whether a pod's status is "running" as a way of checking whether it is alive and well. However, in the reboot test, I went a bit further. I also checked whether the pod's "ready" condition is "true." (To explain, pods have an array of "conditions," one of which is the "ready" condition.) It turns out most other tests don't check the "ready" condition, so the reboot test caught a case where a pod's "ready" condition was "false" when it should have been "true." This is the `fluentd-cloud-logging-<node>` pods.

The annoying thing is that the pods are permanently "not ready" (i.e. they have a "ready" condition that's marked "false"), so we can't just increase the timeout for this check. The pods are also marked with a "true" "ready" condition at the start of the e2e tests. That means they become "not ready" (though still "running") at some point during the e2e tests, usually before the reboot test runs.

I hope these couple paragraphs, and my [comment above](https://github.com/GoogleCloudPlatform/kubernetes/issues/8378#issuecomment-103129874) help clarify what's going on. But as I wrote above, I'm also happy to chat tomorrow.

**Comment by satnam6502** [comment]
Why are these (static/mirror) pods not ready?

S

**Comment by mbforbes** [comment]
That is a great question. I have no idea.

Here's a thought: in #8442, @quinton-hoole expressed concern that we were adding a check with a long timeout to run before every test. I clarified that it would run only as the e2e test suite begins. However, thinking more about this, if we check that all pods in the cluster are "running" and "ready" before every e2e test, we would be able to catch exactly when any pods become "not ready."

**Comment by satnam6502** [comment]
@dchen1107 know everything about the wicked ways of pod status.

**Comment by mbforbes** [comment]
Ping thread participants--if there's anyone that can debug the underlying issue more effectively (I've got a few hammers I'll start trying later today), we might want to rope them in so we can get back to green on GKE CI. 

We can also turn off the reboot test for GKE, though, for those joining in now, this isn't an error in the reboot test itself, but a bug it's finding (see above...)

**Comment by satnam6502** [comment]
I am utterly swamped with other things.

**Comment by quinton-hoole** [comment]
No trying hammers please.   Lets debug this properly.  Just mark the test pending if you want green CI results before then.  I don't personally have time to debug today.

**Comment by mbforbes** [comment]
(I meant debugging using hammers rather than submitting them :-))

**Comment by mbforbes** [comment]
@cjcullen is also having this problem, with the same pods, when bringing up a cluster a few seconds before he starts running e2e tests. A `fluentd-cloud-logging-<node>` pod will first report that it's ready, then exit 0 after some (?) time. With #8442, this can prevent the tests from even running. (I defer to CJ for more details if he so desires.)

**Comment by cjcullen** [comment]
`docker ps -a` on 1 of my 3 nodes suggests that its fluentd container is happily exiting(0) after ~3 seconds of execution.
```
CONTAINER ID        IMAGE                                            COMMAND                CREATED              STATUS                          PORTS               NAMES
10cb1a894622        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   7 seconds ago        Exited (0) 4 seconds ago                            k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_b2269cd8               
caa1c4c0b01a        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   17 seconds ago       Exited (0) 14 seconds ago                           k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_47576a95               
eba1e880129a        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   27 seconds ago       Exited (0) 24 seconds ago                           k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_025203ad               
6428be53a9a7        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   37 seconds ago       Exited (0) 35 seconds ago                           k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_8189506b               
6fd4214b0246        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   47 seconds ago       Exited (0) 45 seconds ago                           k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_9bf16b59               
31cc146aeb6f        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   57 seconds ago       Exited (0) 55 seconds ago                           k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_ef38ecc4               
f2a2055af44d        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   About a minute ago   Exited (0) About a minute ago                       k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_45dcd035               
4a5b4ef17106        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   About a minute ago   Exited (0) About a minute ago                       k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_2ff963df               
0050a37c703a        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   About a minute ago   Exited (0) About a minute ago                       k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_57b037e3               
21f4be10f841        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   About a minute ago   Exited (0) About a minute ago                       k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_622d4245               
3766cfca7d50        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   About a minute ago   Exited (0) About a minute ago                       k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_ec5cea0a               
845c6ec03d94        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   About a minute ago   Exited (0) About a minute ago                       k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_993e7e81               
8a85e7f6d513        gcr.io/google_containers/fluentd-gcp:1.5         "\"/bin/sh -c '/usr/   2 minutes ago        Exited (0) 2 minutes ago                            k8s_fluentd-cloud-logging.b73471a5_fluentd-cloud-logging-k8s-may20-3-node-3_default_16757d3759b931dce16b54e3900bc551_38a8fd93
```
The other 2 nodes have normally running fluentd pods.

**Comment by satnam6502** [comment]
Can you ssh into one of the nodes (I know, could be tricky), then docker exec into the Fluentd container, and then email me the contents of /var/log/google-flunentd.log ? Thanks.
I have a cluster up using gcr.io/google_containers/fluentd-gcp:1.6 which seems pretty stable.
```
# docker ps
CONTAINER ID        IMAGE                                      COMMAND                CREATED             STATUS 
             PORTS               NAMES
45241700a363        gcr.io/google_containers/fluentd-gcp:1.6   "\"/bin/sh -c '/usr/   19 hours ago        Up 19 hours                             k8s_fluentd-cloud-logging.bac571a6_fluentd-cloud-logging-kubernetes-minion-paem_default_83a32f3e71af9530583bdb3ca87bc862_a9c8eaa2   
5f5e9f39e0cb        gcr.io/google_containers/pause:0.8.0       "/pause"               19 hours ago        Up 19hours                             k8s_POD.e4cc795_fluentd-cloud-logging-kubernetes-minion-paem_default_83a32f3e71af9530583bdb3ca87bc862_ce2d4761                    
```

**Comment by satnam6502** [comment]
In case it helps, I think :1.5 used to work ... but stopped working here because of an issue with DNS? :1.6 does not have a dependency on DNS for starting up. This might be the root cause of the issue -- but I am not sure.

**Comment by cjcullen** [comment]
The logs appear empty. I am not able to keep open an interactive shell because the container keeps exiting.

I ended up running `docker exec -it $(docker ps -a | grep fluentd | head -1 | awk '{print $1;}') cat /var/log/google-fluentd.log` until I landed in the window where the container is alive:
```
root@k8s-may20-3-node-3:/home/cjcullen# docker exec -it $(docker ps -a | grep fluentd | head -1 | awk '{print $1;}') cat /var/log/google-fluentd.log
FATA[0000] Error response from daemon: Container 4113f8dd2693 is not running 
root@k8s-may20-3-node-3:/home/cjcullen# docker exec -it $(docker ps -a | grep fluentd | head -1 | awk '{print $1;}') cat /var/log/google-fluentd.log
FATA[0000] Error response from daemon: Container 4113f8dd2693 is not running 
root@k8s-may20-3-node-3:/home/cjcullen# docker exec -it $(docker ps -a | grep fluentd | head -1 | awk '{print $1;}') cat /var/log/google-fluentd.log
FATA[0000] Error response from daemon: Container 4113f8dd2693 is not running 
root@k8s-may20-3-node-3:/home/cjcullen# docker exec -it $(docker ps -a | grep fluentd | head -1 | awk '{print $1;}') cat /var/log/google-fluentd.log
FATA[0000] Error response from daemon: Container 4113f8dd2693 is not running 
root@k8s-may20-3-node-3:/home/cjcullen# docker exec -it $(docker ps -a | grep fluentd | head -1 | awk '{print $1;}') cat /var/log/google-fluentd.log
root@k8s-may20-3-node-3:/home/cjcullen# docker exec -it $(docker ps -a | grep fluentd | head -1 | awk '{print $1;}') cat /var/log/google-fluentd.log
FATA[0000] Error response from daemon: Container 48eb9b501b99 is not running 
```
I figured maybe I'm hitting it before the container has actually written anything to logs, so I switched to `tail -f`but I got similar results:
```
root@k8s-may20-3-node-3:/home/cjcullen# docker exec -it $(docker ps -a | grep fluentd | head -1 | awk '{print $1;}') tail -f /var/log/google-fluentd.log
FATA[0000] Error response from daemon: Container de8b3540f331 is not running 
root@k8s-may20-3-node-3:/home/cjcullen# docker exec -it $(docker ps -a | grep fluentd | head -1 | awk '{print $1;}') tail -f /var/log/google-fluentd.log
root@k8s-may20-3-node-3:/home/cjcullen# 
```

**Comment by jimmidyson** [comment]
We need to change this image to log to stdout/stderr unless there's a good reason not to? I have a PR for this & other fixes coming soon.

**Comment by satnam6502** [comment]
I prefer to have Fluentd log to a log file with -qq

**Comment by cjcullen** [comment]
I updated `/etc/kubernetes/manifests/fluentd-gcp.yaml` to use fluentd-gcp:1.6 and it appears to work.

**Comment by satnam6502** [comment]
So that would point to a DNS failure being the root of the issue -- I think.

S

**Comment by mbforbes** [comment]
Well, our startup e2e check has max no. pods counting DOWN, which seems weird and wrong (though it lets the e2es start)...

```
INFO: 6 / 7 pods in namespace 'default' are running and ready (246.028955ms elapsed)
INFO: 6 / 7 pods in namespace 'default' are running and ready (5.339530088s elapsed)
INFO: 6 / 7 pods in namespace 'default' are running and ready (10.436248982s elapsed)
INFO: 6 / 8 pods in namespace 'default' are running and ready (15.53877564s elapsed)
INFO: 6 / 7 pods in namespace 'default' are running and ready (20.631344115s elapsed)
INFO: 6 / 6 pods in namespace 'default' are running and ready (25.731462049s elapsed)
```

Also, my new test that ensures pods are running and ready before starting can't begin to run because of another pod failure:

```
INFO: Pod monitoring-influx-grafana-controller-q9tk9 failed to be running and ready.
```

ugh...  I'm considering writing crappier tests just so I can get them to run...

**Comment by mbforbes** [comment]
I'm going to add some more logging so we can debug this further.

**Comment by dchen1107** [comment]
Yes, I ran into the same issue caused by #8442. We have so many tests run together. If one of addons service pods run into issues, the rest of tests will be delayed and eventually timeout. Especially both monitoring pod and logging pod are pretty problematically.

**Comment by mbforbes** [comment]
It looks like the underlying issues with pods not being running & ready is gone because this test is passing. 

I'll close this for now.

Fortunately, unless all the pods are being moved out of the default namespace and this isn't looking at them, the reboot test should catch a regression in this problem.
